zcash_client_backend/src/fees/common.rs

```diff
@@ -400,91 +400,74 @@ where
+                    // proposed change, the split policy should recommend only a single change output.
+                    (total_in - total_out_plus_max_fee).unwrap_or(Zatoshis::ZERO),
                 )
-                .unwrap(),
-            );
+            }));
 
             // If we don't have as many change outputs as we expected, recompute the fee.
-            let (fee_with_change, excess_fee) =
-                if usize::from(split_count) < target_change_counts.total_shielded() {
-                    let new_fee_with_change = cfg
-                        .fee_rule
-                        .fee_required(
-                            cfg.params,
-                            target_height,
-                            transparent_input_sizes,
-                            transparent_output_sizes,
-                            sapling_input_count,
-                            sapling_output_count(if change_pool == ShieldedProtocol::Sapling {
-                                usize::from(split_count)
-                            } else {
-                                0
-                            })?,
-                            orchard_action_count(if change_pool == ShieldedProtocol::Orchard {
-                                usize::from(split_count)
-                            } else {
-                                0
-                            })?,
-                        )
-                        .map_err(|fee_error| ChangeError::StrategyError(E::from(fee_error)))?;
-                    (
-                        new_fee_with_change,
-                        (fee_with_change - new_fee_with_change).unwrap_or(Zatoshis::ZERO),
+            let total_fee = if split_count < target_change_counts.total_shielded() {
+                cfg.fee_rule
+                    .fee_required(
+                        cfg.params,
+                        target_height,
+                        transparent_input_sizes,
+                        transparent_output_sizes,
+                        sapling_input_count,
+                        sapling_output_count(if change_pool == ShieldedProtocol::Sapling {
+                            split_count
+                        } else {
+                            0
+                        })?,
+                        orchard_action_count(if change_pool == ShieldedProtocol::Orchard {
+                            split_count
+                        } else {
+                            0
+                        })?,
                     )
-                } else {
-                    (fee_with_change, Zatoshis::ZERO)
-                };
+                    .map_err(|fee_error| ChangeError::StrategyError(E::from(fee_error)))
+            } else {
+                Ok(max_fee)
+            }?;
 
+            let total_change = (total_in - (total_out + total_fee).ok_or_else(overflow)?)
+                .unwrap_or(Zatoshis::ZERO);
```

Comment:
Previously we checked for insufficient funds at line 418. Now we must do it here.

Reply:
I think that the insufficient funds check on (new) line 428 is now sufficient? That should have checked the case where there are insufficient funds to satisfy the minimum fee case. I guess though that if the change strategy has a bug and picks an invalid number of splits, we would want to catch it here. Given that, maybe instead of this being an `unwrap_or`, it should just be `ok_or` and produce an error.